### ISSUE TITLE
Diagnostics process not properly killed

### ISSUE THREAD

**Issue by meldafert** [phpactor/phpactor#2307]
I am using phpactor (2023.06.17-2) as LSP with neovim 0.9.1 (on Ubuntu 22.04, using PHP 8.1)
In a relatively complex legacy project I am working on, the diagnostics process can take a very long time, due to unrelated issues (a misconfiguration on my side).
As expected, phpactor tries to kill the diagnostics process, and logs the message:
```
[2023-06-19 14:32:24] phpactor.WARNING: Killed diagnostics process "1001" because it lived longer than 5s {"channel":"LSP"} []
```
This is good, as otherwise the diagnostics process would hang for too long anyways.

However, it does not actually kill the process.
When spawning the diagnostics process, it creates 3 processes:
1. A shell script wrapper to get the PID of the PHP process (PID 1000)
Command: `sh -c { ('/home/me/source/phpactor/bin/phpactor' 'language-server:diagnostics' '--uri=file:///path/to/file.php') <&3 3<&- 3>/dev/null & } 3<&0; trap "" INT TERM QUIT HUP;pid=$!; echo $pid >&3; wait $pid; RC=$?; echo $RC >&3; exit $RC`
2. A php process (PID 1001)
Command: `php /home/md/source/phpactor/bin/phpactor language-server:diagnostics --uri=file:///path/to/file.php`
3. Another php process that actually runs the diagnostics (PID 1002)
Command: `/usr/bin/php8.1 -n -c /tmp/bxtTPD /home/md/source/phpactor/bin/phpactor language-server:diagnostics --uri=file:///path/to/file.php`

When phpactor kills the process, it kills the process with PID 1001, which also makes the shell wrapper (PID 1000) exit.
However, the process with PID 1002 gets orphaned and continues running in the background.

I have `language_server.diagnostics_on_save: true` set, which means that on every save, a new diagnostics process gets spawned, orphaned, not properly killed, and continues running in the background. This can even lead to an OOM, if I don't notice the issue early enough.

I presume that this can be reproduced with some strategically-placed `sleep(100);` that makes the diagnostics process hang for long enough to trigger the timeout.

It would be good if such a misconfiguration/hang would not lead to orphaned processes, but actually kill the diagnostics process as intended.

**Comment by dantleech** [comment]
hmm, this does work for me. do you have xdebug enabled?

**Comment by meldafert** [comment]
Yes, it seems that was the problem. Should I report this issue then to xdebug-handler instead?

**Comment by meldafert** [comment]
Not sure if this isn't the same issue as https://github.com/composer/xdebug-handler/issues/86, in which case I am not sure if there is much to be done...
Interestingly, `phpactor status` reports to me that xdebug is disabled, presumably because the check happens after the handler restarted it with xdebug disabled.

**Comment by dantleech** [comment]
I think the resolution of that issue may be outdated, the current code does not use `passthru` but AFAICS it still does not forward signals, the linked to code for swoole-bundle has this:

```
    public function forwardSignals(Process $process): void
    {
        \pcntl_async_signals(true);

        $signalForwarder = function (int $signalNo) use ($process): void {
            $process->signal($signalNo);
        };

        foreach (\array_keys(self::SIGNALS_MAP) as $signalNo) {
            \pcntl_signal($signalNo, $signalForwarder);
        }
    }```

i think we can override the `XdebugHandler`:

https://github.com/composer/xdebug-handler/blob/main/src/XdebugHandler.php#L298

and forward the signals to the restarted process
